packages/flutter/lib/src/foundation/timeline.dart

```diff
@@ -0,0 +1,395 @@
+///
+/// It is common for the same block of code to be executed multiple times within
+/// a frame. It is useful to combine multiple executions and report the total
+/// amount of time attributed to that block of code.
+@immutable
+final class AggregatedTimedBlock {
+  /// Creates a timed block of code from a [name] and [duration].
+  ///
+  /// The [name] should be sufficiently unique and descriptive for someone to
+  /// easily tell which part of code was measured.
+  const AggregatedTimedBlock({
+    required this.name,
+    required this.duration,
+    required this.count,
+  }) : assert(duration >= 0);
+
+  /// A readable label for a block of code that was measured.
+  ///
+  /// This field should be sufficiently unique and descriptive for someone to
+  /// easily tell which part of code was measured.
+  final String name;
+
+  /// The sum of [TimedBlock.duration] values of aggretaged blocks.
+  final double duration;
+
+  /// The number of [TimedBlock] objects aggregated.
+  final int count;
+
+  @override
+  String toString() {
+    return 'AggregatedTimedBlock($name, $duration, $count)';
+  }
+}
+
+const int _kSliceSize = 500;
+
+/// A growable list of float64 values with predictable [add] performance.
+///
+/// The list is organized into a "chain" of [Float64List]s. The object starts
+/// with a `Float64List` "slice". When [add] is called, the value is added to
+/// the slice. Once the slice is full, it is moved into the chain, and a new
+/// slice is allocated. Slice size is static and therefore its allocation has
+/// predictable cost. This is unlike the default [List] implementation, which,
+/// when full, doubles its buffer size and copies all old elements into the new
+/// buffer, leading to unpredictable performance. This makes it a poor choice
+/// for recording performance because buffer reallocation would affect the
+/// runtime.
+///
+/// The trade-off is that reading values back from the chain is more expensive
+/// compared to [List] because it requires iterating over multiple slices. This
+/// is a reasonable trade-off for performance metrics, because it is more
+/// important to minimize the overhead while recording metrics, than it is when
+/// reading them.
+final class _Float64ListChain {
+  _Float64ListChain();
+
+  final List<Float64List> _chain = <Float64List>[];
+  Float64List _slice = Float64List(_kSliceSize);
+  int _pointer = 0;
+
+  int get length => _length;
+  int _length = 0;
+
+  /// Adds and [element] to this chain.
+  void add(double element) {
+    _slice[_pointer] = element;
+    _pointer += 1;
+    _length += 1;
+    if (_pointer >= _kSliceSize) {
+      _chain.add(_slice);
+      _slice = Float64List(_kSliceSize);
+      _pointer = 0;
+    }
+  }
+
+  /// Returns all elements added to this chain.
+  ///
+  /// This getter is not optimized to be fast. It is assumed that when metrics
+  /// are read back, they do not affect the timings of the work being
+  /// benchmarked.
+  List<double> get elements {
```

Comment:
Maybe make this a method so it doesn't "feel fast"?

https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#getters-feel-faster-than-methods